tests/test_util_modules.py

```diff
@@ -1,129 +1,304 @@
         await asyncio.sleep(0.1)
-        f2.set_result((123, 'ololo'))
+        f1.set_result((123, 'ololo'))
+        f2.set_result((456, 'pyshpysh'))
+
+    asyncio.create_task(wait_and_assign_async())
+
+    assert (await asyncio.gather(f1, f2)) == [(123, 'ololo'), (456, 'pyshpysh')]
+
+    # await result from separate processes
+    f1, f2 = hivemind.MPFuture(), hivemind.MPFuture()
+
+    def wait_and_assign(future, value):
+        time.sleep(0.1 * random.random())
+        future.set_result(value)
 
-    asyncio.create_task(wait_and_assign())
-    for future in [f1, f2]:
-        res = await future
-        assert res == (123, 'ololo')
+    p1 = mp.Process(target=wait_and_assign, args=(f1, 'abc'))
+    p2 = mp.Process(target=wait_and_assign, args=(f2, 'def'))
+    for p in p1, p2:
+        p.start()
+
+    assert (await asyncio.gather(f1, f2)) == ['abc', 'def']
+    for p in p1, p2:
+        p.join()
 
     # await cancel
-    f1, f2 = hivemind.MPFuture.make_pair()
+    f1, f2 = hivemind.MPFuture(), hivemind.MPFuture()
 
-    async def wait_and_cancel():
-        await asyncio.sleep(0.1)
+    def wait_and_cancel():
+        time.sleep(0.01)
+        f2.set_result(123456)
+        time.sleep(0.1)
         f1.cancel()
 
-    asyncio.create_task(wait_and_cancel())
-    for future in [f1, f2]:
-        with pytest.raises(CancelledError):
-            await future
+    p = mp.Process(target=wait_and_cancel)
+    p.start()
+
+    with pytest.raises(asyncio.CancelledError):
+        # note: it is intended that MPFuture raises Cancel
+        await asyncio.gather(f1, f2)
+
+    p.join()
 
     # await exception
-    f1, f2 = hivemind.MPFuture.make_pair()
+    f1, f2 = hivemind.MPFuture(), hivemind.MPFuture()
 
-    async def wait_and_raise():
-        await asyncio.sleep(0.1)
-        f1.set_exception(SystemError())
+    def wait_and_raise():
+        time.sleep(0.01)
+        f2.set_result(123456)
+        time.sleep(0.1)
+        f1.set_exception(ValueError('we messed up'))
+
+    p = mp.Process(target=wait_and_raise)
+    p.start()
+
+    with pytest.raises(ValueError):
+        # note: it is intended that MPFuture raises Cancel
+        await asyncio.gather(f1, f2)
+
+    p.join()
+
+
+def test_mpfuture_bidirectional():
+    evt = mp.Event()
+    future_from_main = hivemind.MPFuture()
+
+    def _future_creator():
+        future_from_fork = hivemind.MPFuture()
+        future_from_main.set_result(('abc', future_from_fork))
+
+        if future_from_fork.result() == ['we', 'need', 'to', 'go', 'deeper']:
+            evt.set()
+
+    p = mp.Process(target=_future_creator)
+    p.start()
+
+    out = future_from_main.result()
+    assert isinstance(out[1], hivemind.MPFuture)
+    out[1].set_result(['we', 'need', 'to', 'go', 'deeper'])
+
+    p.join()
+    assert evt.is_set()
+
+
+def test_mpfuture_done_callback():
+    receiver, sender = mp.Pipe(duplex=False)
+    evt1, evt2, evt3, evt4, evt5 = (mp.Event() for _ in range(5))
```

Comment:
Maybe just create a list of events and just refer to events[0]?

Reply:
updated the code